packages/validator/src/capability.js

```diff
@@ -635,29 +635,52 @@ class AndMatch {
 }
 
 /**
- * Parses capability `source` using a provided capability `parser`. By default
- * invocation parsing occurs, which respects a capability schema, failing if
- * any non-optional field is missing. If `optional` argument is `true` it will
- * parse capability as delegation, in this case all `nb` fields are considered
- * optional.
+ * @template {string} T
+ * @param {T} match
+ * @param {string} pattern
+ * @returns {T|null}
+ */
+const matchAbility = (match, pattern) =>
+  match === pattern
+    ? match
+    : pattern === '*'
+    ? match
+    : pattern.endsWith('/*') && match.startsWith(pattern.slice(0, -1))
```

Comment:
> If I'm reading this right it seems like this will make it so foo/* always derives foo/bar by default.
>
> Is that part of the [ucan spec](https://github.com/ucan-wg/spec)? In there I only see special processing rules for can=* and can=ucan/* but not anything about any sort general {namespace}/* validation logic.

Correct UCAN spec prescribes no logic no how capabilities must be compared against one another and leaves it up application domain.

> I worry that adding that behavior into ucanto could, at worst, lead a programmer to accidentally authorize things in ucanto they intended to be unauthorized according to ucan spec.

I think this is a moot point, ucanto comes with other various opinions regarding UCANs, e.g. it considers any delegation for the resource matching DID of the issuer to be valid. The UCAN spec does not imply this, yet ucanto chooses to codify it.

Another example is special handling of `./update` capability. UCAN spec does not recognize such capability to be special in any way, yet ucanto validator chooses to recognize it and treat it in a special way.

I consider treatment of `ns/*` just the same, or more specifically ucanto validator would consider such delegations valid without having to manually thread logic in all the capability definitions. That said ucanto can be used with a different validator or no validator at all.

> I worry that adding that behavior into ucanto could, at worst, lead a programmer to accidentally authorize things in ucanto they intended to be unauthorized according to ucan spec.

I think we defend against this with more documentation, reviews tests etc... Given that UCAN spec does not define at all how capabilities are compared the risk of misusing validator is always there. 

Here we just codify how we already were thinking about capabilities, as a matter of fact there several cases where bugs were deployed because we assumed that wildcard patters worked this way while they were not e.g. see https://github.com/web3-storage/ucanto/issues/87#issuecomment-1275225435

> so e.g. I kinda liked how previously you had to explicitly derive in your parsers from {namespace}/* and e.g. it was possible/default to make a {namespace}/foo that was not authorized by {namespace}/*.

That only covered the bottom two in the below image, the chain below would fail to validate when it reached the top one, which is the problem because that is kind of arrangement you'll have with access protocol as described in the pr summary.

```
{ can: "store/*", with: "did:*", size: 100 }
                ^
                |
{ can: "store/*", with: "did:*" }
                ^
                |
{ can:"store/add", with:"did:key:zAlice", size:200 }
```

> @Gozala do you intend to specify that /*-suffix auto-derives logic in [ucan-wg](https://github.com/ucan-wg/spec) ?

Given that UCAN spec intentionally does not define any logic on how capabilities out to be compared it would make much sense IMO. Perhaps years from now there will be desire to have bit more structure around this, in that case I might attempt such a thing.

Reply:
> That said ucanto can be used with a different validator or no validator at all.

good point